python/dgl/dataloading2/graph_storage/csc_sampling_graph.py

```diff
@@ -0,0 +1,309 @@
+    >>> num_nodes = 3
+    >>> ntypes, etypes = ['n1', 'n2'], ['e1', 'e2']
+    >>> hetero_info = (ntypes, etypes, torch.tensor([0, 2]), torch.tensor([0, 1, 0, 1, 1, 0, 0]))
+    >>> graph = from_csc(csc_indptr, indices, num_nodes, hetero_info)
+    >>> print(graph)
+    CSCSamplingGraph(csc_indptr=tensor([0, 2, 5, 7]),
+                    indices=tensor([1, 3, 0, 1, 2, 0, 3]),
+                    num_nodes=3, num_edges=7, is_heterogeneous=True)
+    """
+    if num_nodes is None:
+        num_nodes = csc_indptr.shape[0] - 1
+
+    assert csc_indptr.shape[0] == num_nodes + 1
+    assert torch.max(indices) < num_nodes
+    if hetero_info is None:
+        return CSCSamplingGraph(
+            torch.ops.graphbolt.from_csc(num_nodes, csc_indptr, indices)
+        )
+    else:
+        if etype_sorted is False:
+            ntypes, etypes, node_type_offset, type_per_edge = hetero_info
+            # sort neighbors by edge type id for each node
+            sorted_etype = []
+            sorted_indices = []
+            for s, e in zip(csc_indptr[:-1], csc_indptr[1:]):
+                neighbor_etype, index = torch.sort(type_per_edge[s:e])
+                sorted_etype.append(neighbor_etype)
+                sorted_indices.append(indices[s:e][index])
+            type_per_edge = torch.cat(sorted_etype, dim=0)
+            indices = torch.cat(sorted_indices, dim=0)
+            hetero_info = (ntypes, etypes, node_type_offset, type_per_edge)
+
+        return CSCSamplingGraph(
+            torch.ops.graphbolt.from_csc_with_hetero_info(
+                num_nodes, csc_indptr, indices, *hetero_info
+            )
+        )
+
+
+def from_coo(
+    coo: torch.Tensor,
+    num_nodes: Optional[int] = None,
+    hetero_info: Optional[HeteroInfo] = None,
+) -> CSCSamplingGraph:
+    """
+    Create a CSCSamplingGraph object from a COO representation.
+
+    Parameters
+    ----------
+    coo : torch.Tensor
+        COO format graph.
+    num_nodes : int, optional
+        Number of nodes in the graph, by default None.
+    hetero_info : Optional[HeteroInfo], optional
+        Heterogeneous graph metadata, by default None. Indicates whether the graph is
+        homogeneous or heterogeneous.
+
+    Returns
+    -------
+    CSCSamplingGraph
+        The created CSCSamplingGraph object.
+
+    Examples
+    --------
+    >>> coo = torch.tensor([[0, 1, 2, 2, 3, 3],
+    ...                     [1, 2, 0, 3, 1, 3]])
+    >>> ntypes, etypes = ['n1', 'n2', 'n3'], ['e1', 'e2']
+    >>> hetero_info = (ntypes, etypes, torch.tensor([0, 2, 5]), torch.tensor([0, 1, 0, 1, 0, 1]))
+    >>> graph = dataloading2.from_coo(coo, hetero_info)
+    >>> print(graph)
+    CSCSamplingGraph(csc_indptr=tensor([0, 1, 3, 4, 6]),
+                    indices=tensor([2, 0, 3, 1, 2, 3]),
+                    num_nodes=4, num_edges=6, is_heterogeneous=True)
+    """
+    assert coo.dim() == 2
+
+    if num_nodes is None:
+        num_nodes = torch.max(coo) + (1 if coo.size(1) > 0 else 0)
```

Comment:
Again, be careful for all these checks. Remember this class is designed for large graphs.